docs/architecture/adr-102-rpc-companion.md

```diff
@@ -0,0 +1,341 @@
+from the server.
+
+The **Ingest Service** can control the pruning on the full node via a mechanism to [pruning service](https://github.com/cometbft/cometbft/blob/thane/adr-084-data-companion-pull-api/docs/architecture/adr-101-data-companion-pull-api.md#pruning-service).
+Once the Ingest Service pulls the data from the full node and is able to process it, and it gets an acknowledgement from the database that the data was inserted,
+the **Ingest Service** can communicate with the full node notifying it that a specific height has been processed and set the processed
+height as the `retain height` on the full node signaling this way to the node that this height can be pruned.
+
+If the **Ingest Service** becomes unavailable (e.g. stops), then it should resume synchronization with the full node when it is back online.
+The **Ingest Service** should query the full node for the last `retain height` and the **Ingest Service** should request
+and process all the heights missing on the database until it catches up with the full node latest height.
+
+In case the **Ingest Service** becomes unavailable for a long time and there are several heights to be synchronized, it is
+important for the **Ingest Service** to do it in a throttled way in order not to stress the full node and cause issues in its consensus processing.
+
+### [Database](#database)
+
+The database stores the data retrieved from the full node and provide this data for the RPC server instance. Since the frequency
+that blocks are generated on the chain are in the range from 5 seconds to 7 seconds on average, the _write_ back pressure is not
+very high from a modern database perspective. While the frequency and number of requests for reading the data from the database will
+be much larger due to the fact that the RPC service instance can be scaled. Therefore, a database that provides a high read
+throughput should be favored.
+
+For this is initial solution implementation it is proposed that the relational database [Postgresql](https://www.postgresql.org/) should be used in order to support
+the RPC scalability and this will also provide more flexibility when implementing the RPC v2 endpoints that can return data
+in different forms and database indexes might also be leveraged in order to boost the query responses performance.
+
+The data needs to be available both for the Ingest Service (writes) and the RPC server instance (reads) and these service might be running from different machines so an embedded database
+is not recommended in this case since accessing the data remotely might not be optimal for an embedded key-value database. Also
+since the RPC might have many server instances (or processes) running that will need to retrieve data concurrently it is recommended to use
+a well-known robust database engine that can support such a load.
+
+Also, a database that can support ACID transactions is important to provide more guarantees that
```

Comment:
Actually, I think the most important point of having ACID transactions is to ensure that clients (RPC) don't read incomplete data, namely data that was only partially written to the database. Any database provides write guarantees, but if the write is made of multiple steps, readers can access incomplete data.